src/main/kotlin/com/manimdsl/ManimParserVisitor.kt

```diff
@@ -7,9 +7,66 @@ import com.manimdsl.frontend.*
 class ManimParserVisitor : ManimParserBaseVisitor<ASTNode>() {
     val symbolTable = SymbolTableVisitor()
     private val semanticAnalyser = SemanticAnalysis()
+    private var inFunction: Boolean = false
+    private var functionReturnType: Type = VoidType
 
     override fun visitProgram(ctx: ManimParser.ProgramContext): ProgramNode {
-        return ProgramNode(ctx.stat().map { visit(it) as StatementNode })
+        return ProgramNode(
+                ctx.function().map { visit(it) as FunctionNode },
+                ctx.stat().map { visit(it) as StatementNode })
+    }
+
+    override fun visitFunction(ctx: ManimParser.FunctionContext): FunctionNode {
+        inFunction = true
+        val identifier = ctx.IDENT().symbol.text
+        semanticAnalyser.redeclaredVariableCheck(symbolTable, identifier, ctx)
+
+        val type = if (ctx.type() != null) {
+            visit(ctx.type()) as Type
+        } else {
+            VoidType
+        }
+        functionReturnType = type
+
+        val scope = symbolTable.enterScope()
+        val parameters: List<ParameterNode> =
+            visitParameterList(ctx.param_list() as ManimParser.ParameterListContext?).parameters
+        val statements = ctx.stat().map { visit(it) as StatementNode }
```

Comment:
We'll need a null check on `ctx.stat()` now right? Now that statements are optional

Reply:
I think we just get an empty list if the function body is empty